File: app/src/main/java/com/travlog/android/apps/ui/viewholders/NoteViewHolder.java
package com.travlog.android.apps.ui.viewholders;

import android.support.annotation.NonNull;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.travlog.android.apps.R;
import com.travlog.android.apps.models.Note;

import butterknife.BindView;
import butterknife.ButterKnife;

public class NoteViewHolder extends RecyclerView.ViewHolder {

    public interface Delegate {
        void noteViewHolderItemClick(@NonNull NoteViewHolder viewHolder, @NonNull Note note);
    }

    private final Delegate delegate;

    @BindView(R.id.title)
    AppCompatTextView title;

    public NoteViewHolder(final @NonNull View itemView, final @NonNull Delegate delegate) {
        super(itemView);
        ButterKnife.bind(this, itemView);
        this.delegate = delegate;
    }

    public void bindData(final @NonNull Note item) {
        this.title.setText(item.title);

        itemView.setOnClickListener(view -> this.delegate.noteViewHolderItemClick(this, item));
    }
}

File: app/src/main/java/com/travlog/android/apps/ui/viewholders/NoteViewHolder.kt
package com.travlog.android.apps.ui.viewholders

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.travlog.android.apps.R
import com.travlog.android.apps.libs.utils.DateTimeUtils
import com.travlog.android.apps.models.Note
import kotlinx.android.synthetic.main.i_note.view.*
import org.joda.time.DateTime

class NoteViewHolder(itemView: View, private val delegate: Delegate) : RecyclerView.ViewHolder(itemView) {

    interface Delegate {
        fun noteViewHolderItemClick(viewHolder: NoteViewHolder, note: Note)
    }

    fun bindData(item: Note) {
        this.itemView.note_title.text = item.title

        when {
            item.destinations.isNotEmpty() -> {
                val destination = item.destinations[0]

                this.itemView.note_destination.text = destination?.location?.name.orEmpty()
                this.itemView.note_date.text = itemView.context.getString(R.string.note_date,
                        DateTimeUtils.mediumDate(DateTime(destination?.startDate)),
                        DateTimeUtils.mediumDate(DateTime(destination?.endDate)))
            }
            else -> {
                this.itemView.note_destination.text = ""
                this.itemView.note_date.text = ""
            }
        }

        itemView.setOnClickListener { this.delegate.noteViewHolderItemClick(this, item) }
    }
}
